### ISSUE TITLE
Downloading wrong language

### ISSUE THREAD

**Issue by corbin-lounsbury** [meeb/tubesync#368]
Recently, some videos in a subscribed channel are randomly downloading the Spanish language versions of the videos when I am wanting the English version. There's nothing in the settings that I can see that would make that the default. 

The channel uploads their videos in English as the original and Spanish as dubbed. 

The channel: https://www.youtube.com/@JaredOwen

Videos that are downloaded in Spanish:

https://youtu.be/rKp4pe92ljg

https://youtu.be/0-oQRSViZQE

https://youtu.be/GkI_2yiIb48

**Comment by meeb** [comment]
Thanks, I'd not seen this previously so I would assume I need to add some language selection flag.

**Comment by ryjogo** [comment]
I just realised about 1500 videos of my subscibed channel are not in spanish :( any eta on this ?

**Comment by mckayc** [comment]
I am noticing the same issue as well. Especially with the more popular Youtube channels where they upload multiple languages, it almost never downloads with the English audio. It looks like Youtube-DL has a way to specify the language, so I would think it would be possible to do here as well. https://www.reddit.com/r/youtubedl/comments/114624f/how_to_choose_audio_language_for_youtube_video/

**Comment by davidkylenz** [comment]
The ability to select the language would be nice, I’m starting to see a few channels supporting multiple language audio tracks and it never downloads the English variant.

**Comment by mylogon341** [comment]
Same here. The latest Mark Rober video downloaded the Spanish dub. It would be good to be able to set the default language as a flag somewhere in the UI, ideally on a global level.

**Comment by meeb** [comment]
Thanks for the reports everyone, I'll look into adding this as an option and default to English when I get time this week.

**Comment by bawitdaba** [comment]
> Thanks for the reports everyone, I'll look into adding this as an option and default to English when I get time this week.

I was able to modify utils.py to accomplish this, setting English to default. I added line 170 and lines 186-190. Tested with a Mark Rober  video and it re-selected the correct audio track. This basically ignores any audio track with a - in the ID where the note doesn't contain English.

I also have a block of code in here to do the same with skipping (Premium) resolutions.

![image](https://github.com/meeb/tubesync/assets/12944594/eba85582-0364-4ba7-abca-37076a2ff7c1)

**Comment by irishj** [comment]
Thanks for your code @bawitdaba, worked a treat for me with Mark Rober's videos also :)

Attaching the modified file here for anyone else who wants to use it, without having to manually add the code themselves.

[utils.zip](https://github.com/meeb/tubesync/files/13465159/utils.zip)

The file needs to go into /app/sync

Cheers

**Comment by meeb** [comment]
Thanks! The premium skip looks fine. Defaulting to English with no dynamic option probably isn't sensible for everyone though. Do you want to add a PR for the premium skip section?

**Comment by frob** [comment]
Defaulting to English may not be the best solution but it is likely an 80% solution. 

Looking at the UI for youtube it looks like, maybe, `original` is the correct default option. All the multi-language videos I found marked the language as "English original"

Turning this into a PR would be good. Then work could be done on adding a flag to the UI for downloading a preferred language with original being the fallback.

**Comment by meeb** [comment]
Sounds good. I just meant the defaulting to English shouldn't be in the PR. If you create PR for the rest of it I'd be happy to review it.

**Comment by corbin-lounsbury** [comment]
Hello! Is this still in the pipeline to get worked or has it been resolved? I'm not seeing anything different in my instance

**Comment by jsclifford** [comment]
I would like to see this fixed as well. I can't download any Mark Robber Videos.  The audio language ranges from Russion, to Spanish, and even French.  Is anyone working on this.

**Comment by tack32** [comment]
This is amazing because I was trying to download a Mark Rober video today and even though it played on YT in English, which I want, it downloads in Spanish. I'm actually using replay media catcher and haven't tried the program you're talking about here.

**Comment by djismgaming** [comment]
For now, you can use the file from @irishj post.

If you use `docker compose` you can put the file alongside your `compose.yaml` and reference it like this in it:

```yaml
    volumes:
      - ./utils.py:/app/sync/utils.py
```

**Comment by Catchwa** [comment]
I added in the patch via docker and triggered a re-download of an existing video with the wrong language, but I still get the same issue. Do I have to delete the whole channel rather than an individual video?


```

me@nas:~$ sudo docker exec -it tubesync grep -nC 11 "and not 'ENGLISH'" /app/sync/utils.py
170-    format_id = format_dict.get('format_id', '').strip().upper()
171-    format_full = format_dict.get('format_note', '').strip().upper()
172-    format_str = format_full[:-2] if format_full.endswith('60') else format_full
173-    format_str = format_str.strip()
174-    format_str = format_str[:-3] if format_str.endswith('HDR') else format_str
175-    format_str = format_str.strip()
176-    format_str = format_str[:-2] if format_str.endswith('60') else format_str
177-    format_str = format_str.strip()
178-    is_hls = True
179-    is_dash = False
180-
181:    if '-' in format_id and not 'ENGLISH' in format_full:
182-        vcodec = None
183-        acodec = None
184-        height = 0
185-        width = 0
186-
187-    if 'PREMIUM' in format_full:
188-        vcodec = None
189-        acodec = None
190-        height = 0
191-        width = 0
192-
```

**Comment by ChirpyTurnip** [comment]
I don't know what is causing the problem but it is getting worse. When I first started all my videos were in English, but now I have three or four that are coming though in other languages but never in English....it seems every week or so another stream flips over to foreign. Is the fix implemented.....?

**Comment by superjedi27** [comment]
If anyone else comes across this and is kind of confused like I was-
Place the utils.py file in one of your shares. Then open the terminal for unRAID at the top and run the following command-
sudo docker cp /mnt/user/insertYourShareNameInfo/utils.py TubeSync:/app/sync

For me TubeSync is capitalized cause thats how I named the docker, yours might be all lower case or it could be something completely different. Whatever its named on the Docker tab use that, then restart tubesync and redownload the videos. Also make sure you are using your correct file path for the utils.py file

**Comment by meeb** [comment]
Just to mention this issue is next on my list when I get some time to look at it.

**Comment by ChirpyTurnip** [comment]
That would be awesome! It is rather frustrating how random it seems to be....maybe it grabs the 'latest' audio and the translations are uploaded after the original. 

Appreciate the effort, and no rush - we just grin and bear the ads on YouTube and that will be less pain than coding a solution I'm sure!

**Comment by RichardHyde** [comment]
In the last week I've started getting the occasional video downloaded with non-English language audio (French, German, and Spanish so far, all the videos were originally in English). I tried applying the updated `utils.py` file linked above, but it's still happening.

Let me know if you need any more information to help debug the issue.

**Comment by meeb** [comment]
Thanks @RichardHyde - the issue and cause is known, it's just not been properly fixed yet.

**Comment by jesusdf** [comment]
The automatic AI audio translation is becoming a big issue, since Google is applying it to everything.
I think that a simple fix would be to download the original audio always by default (currently the language is autodetected), and after that, maybe creating a new option to choose a preferred language per channel.

Example:

yt-dlp --list-formats "https://www.youtube.com/watch?v=jmhBS3stViw" | grep original

`233-5 mp4   audio only        |                   m3u8  | audio only           unknown             [en-US] American English - original
234-5 mp4   audio only        |                   m3u8  | audio only           unknown             [en-US] American English - original
249-6 webm  audio only      2 |    6.44MiB    50k https | audio only           opus        50k 48k [en-US] English (United States) original (default), low, THROTTLED, webm_dash
250-6 webm  audio only      2 |    8.02MiB    62k https | audio only           opus        62k 48k [en-US] English (United States) original (default), low, THROTTLED, webm_dash
251-6 webm  audio only      2 |   15.21MiB   118k https | audio only           opus       118k 48k [en-US] English (United States) original (default), medium, THROTTLED, webm_dash
139-6 m4a   audio only      2 |    6.31MiB    49k https | audio only           mp4a.40.5   49k 22k [en-US] English (United States) original (default), low, m4a_dash
140-6 m4a   audio only      2 |   16.74MiB   129k https | audio only           mp4a.40.2  129k 44k [en-US] English (United States) original (default), medium, m4a_dash`

**Comment by arionl** [comment]
Just adding a "me too" here. I've noticed this on 2-3 videos in one specific channel I follow over the past month. Getting auto-generated dubs, usually Spanish, on the videos. I can send links to specific videos if it would be helpful. I'll look into the work-around posted above for now..

**Comment by tcely** [comment]
@meeb I think I fixed this unintentionally in #566 when we stopped badly sorting audio formats

**Comment by meeb** [comment]
Yeah that's quite possible. Can anyone confirm if this is better now?

**Comment by jesusdf** [comment]
I think that it is fixed now, at least the latest videos I've downloaded are in their original language.

**Comment by corbin-lounsbury** [comment]
Hopeful news! My tubesync instance went wonky and I haven't used it for a while. Let me make a new one with the latest image and I'll report back!

**Comment by corbin-lounsbury** [comment]
I started a new instance and it seems to be fixed! At least the same files I had trouble with last time are in english this time.

**Comment by tcely** [comment]
That's great news!

Is this issue ready to be closed?

**Comment by timwhite** [comment]
I got my instance to redownload a channel I was having issues with, and it does seem to be fixed too!

**Comment by meeb** [comment]
Thanks everyone. Unless there are any other reports of this today I'll close this soon.

**Comment by gurdy4me** [comment]
> Thanks everyone. Unless there are any other reports of this today I'll close this soon.

Sorry, but I'm an idiot and I'm still trying to figure out how to get ClipGrab to download youtube videos with their original language track (I require English).  I'm still getting the random AI-generated foreign language audio tracks.  If this problem has been "fixed," can someone please help me figure out how to make it work?

**Comment by tcely** [comment]
> Hmm, now that I look more closely, seems like this isn't the place to discuss ClipGrab at all!

Indeed. This is not [`ClipGrab`](https://github.com/FreedomBen/clipgrab).

However, it does appear that you could replace that software with TubeSync, if you wanted to.

**Comment by gurdy4me** [comment]
> > Hmm, now that I look more closely, seems like this isn't the place to discuss ClipGrab at all!
> 
> Indeed. This is not [`ClipGrab`](https://github.com/FreedomBen/clipgrab).
> 
> However, it does appear that you could replace that software with TubeSync, if you wanted to.

I warned you I was an idiot!  Thank you for the suggestion, I'll check it out...

**Comment by tcely** [comment]
@meeb Please close this when you have a moment.

**Comment by meeb** [comment]
Thanks @tcely ! I'll close this for now.
